src/Mod/Sketcher/Gui/EditModeGridCoinManager.cpp

```diff
@@ -0,0 +1,244 @@
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <Inventor/nodes/SoSeparator.h>
+# include <Inventor/nodes/SoBaseColor.h>
+# include <Inventor/nodes/SoDepthBuffer.h>
+# include <Inventor/nodes/SoDrawStyle.h>
+# include <Inventor/nodes/SoLineSet.h>
+# include <Inventor/nodes/SoMaterial.h>
+# include <Inventor/nodes/SoPickStyle.h>
+# include <Inventor/nodes/SoSeparator.h>
+# include <Inventor/nodes/SoVertexProperty.h>
+
+# include <Inventor/SbVec3f.h>
+#endif  // #ifndef _PreComp_
+
+#include <Gui/Application.h>
+#include <Gui/View3DInventor.h>
+#include <Gui/View3DInventorViewer.h>
+#include <Gui/SoFCBoundingBox.h>
+
+#include "ViewProviderSketch.h"
+#include "EditModeCoinManager.h"
+#include "EditModeGridCoinManager.h"
+
+using namespace SketcherGui;
+
+//**************************** EditModeGridCoinManager class ******************************
+
+EditModeGridCoinManager::EditModeGridCoinManager(   ViewProviderSketch &vp,
+                                                    EditModeScenegraphNodes & editModeScenegraph):
+    camCenterOnSketch(SbVec3f(0., 0., 0.)),
+    camMaxDimension(200.),
+    viewProvider(vp),
+    editModeScenegraphNodes(editModeScenegraph)
+{
+}
+
+EditModeGridCoinManager::~EditModeGridCoinManager()
+{}
+
+double EditModeGridCoinManager::getGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0)
+        return viewProvider.GridSize.getValue();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    int numberOfLines = std::max(pixelWidth, pixelHeight) / hGrp->GetInt("GridSizePixelThreshold", 15);
+    ParameterGrp::handle hGrp2 = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    int units = hGrp2->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feets
+    double unitMultiplier = (units == 2 || units == 3) ? 25.4 : (units == 5 || units == 7) ? 304.8 : 1;
+
+    int subdivisions = hGrp->GetInt("GridNumberSubdivision", 10);
+    double newGridSize = unitMultiplier * pow(subdivisions, 1 + floor(log(camMaxDimension / unitMultiplier / numberOfLines) / log(subdivisions)));
+
+    //cap the grid size
+    newGridSize = std::max(newGridSize, 0.000001);
+    newGridSize = std::min(newGridSize, 10000000.0);
+
+    viewProvider.gridSizeBlocker = true; //TODO: replace with Base::ConnectionBlocker block(??? connection to onChanged);
+    viewProvider.GridSize.setValue(newGridSize); //grid size must be set for grid snap. But we need to block it from calling createGrid.

```

Comment:
OK, after having a deeper I find no actual way to block the call, but I would suggest to remove the `gridSizeBlocker` and instead protect from recursive calls as this:
```suggestion
    if (newGridSize != viewProvider.GridSize.getValue()) // protect from recursive calls
        viewProvider.GridSize.setValue(newGridSize);
```

Reply:
Yes this makes more sens anyway.